### PR TITLE
feat(distribute): send an `Epoch-Distribution-Notice` when epochs are…

### PR DESCRIPTION
**Pull request by dtfiedler**
… distributed

### CODE REVIEW

File: tests/tick.test.mjs [hunk 1]
```diff
@@ -455,8 +477,29 @@ describe('Tick', async () => {
       timestamp: distributionTimestamp,
     });
 
-    // assert no error tag
-    assertNoResultError(distributionTick);
+    // assert multiple messages are sent given the tick notice, epoch created notice and epoch distribution notice
+    assert.equal(distributionTick.result.Messages.length, 4); // cannot explain why this is 4, i'd expect it to be 3 messages (1 tick notice, 1 epoch created notice, 1 epoch distribution notice)
```
Comment (dtfiedler) [inline tests/tick.test.mjs:481]: added a comment here - we are getting one extra `Tick-Notice` at the end of the message array, unclear why. will spend some more time on it

Comment (fedellen) [reply tests/tick.test.mjs:481]: Whats the 4th message?

**Review by fedellen — DISMISSED** [review]

File: src/epochs.lua [hunk 2]
```diff
@@ -525,11 +525,14 @@ function epochs.computeTotalEligibleRewardsForEpoch(epochIndex, prescribedObserv
...
-		totalEligibleRewards * constants.observerRewardRatio / utils.lengthOfTable(prescribedObserversLookup)
-	)
+	local eligibleGatewayReward = #activeGatewayAddresses > 0
+			and math.floor(totalEligibleRewards * constants.gatewayOperatorRewardRatio / #activeGatewayAddresses)
+		or 0
+	local eligibleObserverReward = utils.lengthOfTable(prescribedObserversLookup) > 0
+			and math.floor(
+				totalEligibleRewards * constants.observerRewardRatio / utils.lengthOfTable(prescribedObserversLookup)
+			)
+		or 0
```
Comment (dtfiedler) [inline src/epochs.lua:535]: this was causing `inf` due to rounding by 0 when there are no gateways joined

File: src/main.lua [hunk 3]
```diff
@@ -1701,41 +1714,56 @@ addEventingHandler("distribute", utils.hasMatchingTag("Action", "Tick"), functio
 	local tickedRewardDistributions = {}
 	local totalTickedRewardsDistributed = 0
 
+	print("Ticking from " .. lastTickedEpochIndex .. " to " .. targetCurrentEpochIndex)
+
 	--- tick to the newest epoch, and stub out any epochs that are not yet created
 	for i = lastTickedEpochIndex + 1, targetCurrentEpochIndex do
+		print("Ticking epoch " .. i)
```
Comment (dtfiedler) [inline src/main.lua:1721]: these logs proved useful when debuggin tests, so left them in

File: src/main.lua [hunk 4]
```diff
@@ -1701,41 +1714,56 @@ addEventingHandler("distribute", utils.hasMatchingTag("Action", "Tick"), functio
...
 		-- and that we don't try and distribute before an epoch is created
 		local tickTimestamp = math.min(msg.Timestamp, epochDistributionTimestamp)
 		-- TODO: if we need to "recover" epochs, we can't rely on just the current message hashchain and block height,
 		-- we should set the prescribed observers and names to empty arrays and distribute rewards accordingly
 		local tickResult = tick.tickEpoch(tickTimestamp, blockHeight, hashchain, msgId)
-		if tickTimestamp == epochDistributionTimestamp then
-			-- if we are distributing rewards, we should update the last ticked epoch index to the current epoch index
-			LastTickedEpochIndex = i
-			table.insert(tickedEpochIndexes, i)
-		end
```
Comment (dtfiedler) [inline src/main.lua:1717]: this never made sense - if we've ticked the epoch at all, we should increment the `LastTickedEpochIndex`. We can also consider replacing it with two more granular global variables `LastCreatedEpochIndex` and `LastDistributedEpochIndex`

File: tests/tick.test.mjs [hunk 5]
```diff
@@ -455,8 +474,31 @@ describe('Tick', async () => {
...
+      ).value,
+      'Epoch-Distribution-Notice',
+    );
+
+    assert.equal(
+      distributionTick.result.Messages[2].Tags.find(
+        (tag) => tag.name === 'Action',
+      ).value,
+      'Tick-Notice',
+    );
```
Comment (dtfiedler) [inline tests/tick.test.mjs:501]: validates all three messages get sent when a new epoch is created and distributed

File: tests/tick.test.mjs
Comment (dtfiedler) [inline tests/tick.test.mjs:1]: most of the changes in this file are to leverage the reusable `tick` handler in `helpers.mjs` and update state with correct variables from the contract

File: tests/tick.test.mjs (hunk 1 above)
Comment (dtfiedler) [reply tests/tick.test.mjs]: it was mainly due to test setup, and some minor edge cases in how we create the first epoch. The latest commits have a few changes that address the issue.

File: src/main.lua (hunk 4 above)
Comment (dtfiedler) [reply src/main.lua:1717]: we could also add `LastDemandFactorUpdateTimestamp` to track when the demand factor was last updated.

**Review by fedellen — APPROVED** [review]

File: src/main.lua (hunk 4 above)
Comment (dtfiedler) [reply src/main.lua:1718]: merged into his branch on https://github.com/ar-io/ar-io-network-process/pull/317/commits/49661f138b4e8a7d8a75dfe4cbceb76526853f7c

File: src/main.lua [hunk 6]
```diff
@@ -1667,79 +1681,84 @@ end)
...
-		Send(msg, {
-			Target = msg.From,
-			Action = "Tick-Notice",
-			LastTickedEpochIndex = tostring(LastTickedEpochIndex),
-			Data = json.encode(tickResult),
-		})
+	for i = lastCreatedEpochIndex, targetCurrentEpochIndex do
```
Comment (dtfiedler) [inline src/main.lua:1722]: this logically makes more sense - instead of "going back one epoch", start this process at the last known created epoch and go until we reach the target current epoch
